File: src/test/java/org/meetkt/catalogue/fixture/ProductFixture.java
package org.meetkt.catalogue.fixture;

import org.meetkt.catalogue.domain.model.Product;
import org.meetkt.catalogue.domain.model.ProductId;
import org.meetkt.basket.domain.model.Price;

public class ProductFixture {

    private String barcode = "ZZZZZZZ";
    private ProductId productId;
    private Price price = Price.ZERO;
    private String description = "Default description";

    private ProductFixture() {
    }

    public static ProductFixture aProduct() {
        return new ProductFixture();
    }

    public ProductFixture withBarcode(String barcode) {
        this.barcode = barcode;
        return this;
    }

    public ProductFixture withProductId(String productId) {
        this.productId = new ProductId(productId);
        return this;
    }

    public ProductFixture withPriceInt(int price) {
        this.price = new Price(price);
        return this;
    }

    public ProductFixture withPriceDouble(double price) {
        this.price = new Price(price);
        return this;
    }

    public ProductFixture withDescription(String description) {
        this.description = description;
        return this;
    }

    public Product build() {
        return new Product(productId, barcode, price, description);
    }
}

File: src/test/kotlin/in/meetkt/catalogue/fixture/ProductFixture.kt
package `in`.meetkt.catalogue.fixture

import `in`.meetkt.basket.domain.model.Price
import `in`.meetkt.catalogue.domain.model.Product
import `in`.meetkt.catalogue.domain.model.ProductId

class ProductFixture private constructor() {

    private var barcode = "ZZZZZZZ"
    private var productId: ProductId = ProductId("00000")
    private var price = Price.ZERO

    companion object {
        fun aProduct() = ProductFixture()
    }

    fun build(): Product = Product(productId, barcode, price)

    fun withProductId(id: String): ProductFixture = this.also { it.productId = ProductId(id) }

    fun withBarcode(barcode: String): ProductFixture = this.also { it.barcode = barcode }

    fun withPriceInt(price: Int) = withPriceDouble(price.toDouble())

    fun withPriceDouble(price: Double) = this.also { it.price = Price(price) }
}